server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -513,10 +513,17 @@ private boolean canMatch(LeafReaderContext ctx) throws IOException {
     }
 
     private boolean canMatchSearchAfter(LeafReaderContext ctx) throws IOException {
-        if (searchContext.searchAfter() != null && searchContext.request() != null && searchContext.request().source() != null) {
+        // Skipping search on shard/segment entirely can cause mismatch on total_tracking_hits, hence skip only if
+        // track_total_hits is false.
+        if (searchContext.searchAfter() != null
```

Comment:
@jainankitk The potential improvements are mostly related to short circuit. For example, we can check the conditions upfront here and avoid to build sort field or read min max values unnecessary.